src/core_init_atmosphere/mpas_kd_tree.F

```diff
@@ -0,0 +1,447 @@
+module mpas_kd_tree
+    
+   !***********************************************************************
+   !
+   !  module mpas_kd_tree
+   !
+   !> \brief   MPAS KD-Tree module
+   !> \author  Miles A. Curry 
+   !> \date    03/04/19
+   !> \details
+   !>
+   !>
+   !> TODO: Fail immediatly edge cases  
+   !>       1. Different number of dimensions
+   !>       2. No points
+   !>       3. No dimensions
+   !>
+   !> TODO: Add nn-search - Search for n nearest neighbors instead of only 1
+   !>
+   !
+   !-----------------------------------------------------------------------
+   implicit none
+
+   private
+
+   public :: kdnode
+
+   ! Public Subroutines
+   public :: mpas_kd_insert
+   public :: mpas_kd_construct
+   public :: mpas_kd_search
+   public :: mpas_kd_free
+
+   type kdnode
+      type (kdnode), pointer :: left => null()
+      type (kdnode), pointer :: right => null()
+
+      integer :: split_dim
+      real, dimension(:), pointer :: point
+      
+      integer :: cell
+      logical :: is_halo
+   end type kdnode
```

Comment:
We should update the kdnode to include a payload pointer. Lets update it and then create tests for it.